File: parser/src/lexer.rs
use std::{collections::HashMap, fmt::Write};

use lazy_static::lazy_static;

use super::{Error, Result};

/// Represents the set of reserved keywords.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Keyword {
    If,       // `if`
    Else,     // `else`
    Int,      // `int`
    Unsigned, // `unsigned`
    Void,     // `void`
    Bool,     // `bool `
    Goto,     // `goto`
    Return,   // `return`
}

impl Keyword {
    fn from_str(kw: &str) -> Option<Self> {
        match kw {
            "if" => Some(Self::If),
            "else" => Some(Self::Else),
            "int" => Some(Self::Int),
            "unsigned" => Some(Self::Unsigned),
            "void" => Some(Self::Void),
            "bool" => Some(Self::Bool),
            "goto" => Some(Self::Goto),
            "return" => Some(Self::Return),
            _ => None,
        }
    }
}

/// Represents the set of recognizable tokens.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    Eq,    // '=='
    Gq,    // '>='
    Lq,    // '<='
    Shl,   // `>>`
    Shr,   // `<<`,
    Tilda, // `~`

    LPar, // `(`
    RPar, // `)`
    LSq,  // `[`
    RSq,  // `]`
    LCb,  // `{`
    RCb,  // `}`

    Plus,    // `+`
    Minus,   // `-`
    Star,    // `*`
    Slash,   // `/`
    Percent, // `%`
    Assign,  // `=`

    Gt,  // '>'
    Lt,  // '<'
    And, // '&'
    Xor, // '^'
    Or,  // '|'
    Not, // '!'

    Comma,     // ','
    Semicolon, // `;`
    Colon,     // `:`

    Keyword(Keyword),
    Ident(String), // [a-zA-Z][a-zA-Z0-9]*
    Number(u64),
}

impl Token {
    pub fn into_ident(self) -> Option<String> {
        match self {
            Self::Ident(n) => Some(n),
            _ => None,
        }
    }
}

/// Make a new ident.
fn ident<T: Into<String>>(s: T) -> Token {
    Token::Ident(s.into())
}

/// Macro for tokens that are constructable without possible errors.
macro_rules! primitive_tokens {
    ( $(($regex: literal, $token: expr)),* $(,)? ) => {{
        let v: Vec::<(&'static str, fn(&str) -> Result<Token>)> = vec![
            $(($regex, |_| Ok($token))),*
        ];
        v
    }};
}

lazy_static! {
    // Regexes are matched greedily from the first to the last.
    static ref TOKEN_REGEXS: Vec<(&'static str, fn(&str) -> Result<Token>)> = {
        // Non-primitive tokens.
        let mut regexs: Vec<(&'static str, fn(&str) -> Result<Token>)> = vec![
            (r"[+-]?[0-9]+", |a| parse_number(a).map(Token::Number)),
            (r"[a-zA-Z_][a-zA-Z0-9_]*", |a: &str| {
                Ok(
                    // Firstly try to extract keyword, then make an ident.
                    Keyword::from_str(&a)
                    .map(Token::Keyword)
                    .unwrap_or_else(|| Token::Ident(a.to_owned()) )
                )
            })
        ];
        // Primitive tokens.
        regexs.append(&mut primitive_tokens! {
            (r"==", Token::Eq),
            (r">=", Token::Gq),
            (r"<=", Token::Lq),
            (r"<<", Token::Shl),
            (r">>", Token::Shr),
            (r"~", Token::Tilda),
            (r"\(", Token::LPar),
            (r"\)", Token::RPar),
            (r"\[", Token::LSq),
            (r"\]", Token::RSq),
            (r"\{", Token::LCb),
            (r"\}", Token::RCb),
            (r"\+", Token::Plus),
            (r"\-", Token::Minus),
            (r"\*", Token::Star),
            (r"/", Token::Slash),
            (r"%", Token::Percent),
            (r"=", Token::Assign),
            (r">", Token::Gt),
            (r"<", Token::Lt),
            (r"&", Token::And),
            (r"\^", Token::Xor),
            (r"\|", Token::Or),
            (r"!", Token::Not),
            (r",", Token::Comma),
            (r";", Token::Semicolon),
            (r":", Token::Colon)
        });
        regexs
    };
}

/// Lexer splits an input stream into the tokens.
pub struct Lexer<'a> {
    input: &'a str,
    regexp: regex::Regex,
    /// Token constructors for specific match groups.
    cons: HashMap<usize, &'static fn(&str) -> Result<Token>>,
}

fn parse_number(s: &str) -> Result<u64> {
    s.parse().map_err(Error::ParseIntError)
}

impl<'a> Lexer<'a> {
    /// Make a new [`Lexer`].
    pub fn new(input: &'a str) -> Self {
        let mut cons = HashMap::new();
        let mut groups = String::with_capacity(512);

        // Collect regexs for tokens.
        for (i, (re, con)) in TOKEN_REGEXS.iter().enumerate() {
            write!(&mut groups, "|({})", re).unwrap();
            // Token groups started from 2, first two groups are reserved.
            cons.insert(i + 2, con);
        }

        // Match groups are:
        // * The whole matched.
        // * EOF.
        // * Token specific groups.
        let regexp = regex::Regex::new(&format!(r"^\s*(?:($){})", groups)).unwrap();
        Self {
            input,
            regexp,
            cons,
        }
    }

    /// Get next token from the input.
    /// If EOF is reached, returns [`None`].
    ///
    pub fn next(&mut self) -> Result<Option<Token>> {
        if self.input.is_empty() {
            // EOF.
            Ok(None)
        } else {
            match self.regexp.captures(&self.input) {
                Some(matched) => {
                    // Shift input.
                    // The whole match (0-group) always starts from the beginning.
                    self.input = &self.input[matched.get(0).unwrap().end()..];

                    Ok(if matched.get(1).is_some() {
                        // EOF matched.
                        None
                    } else {
                        for (group_num, cons) in self.cons.iter() {
                            if let Some(matched) = matched.get(*group_num) {
                                return cons(matched.as_str()).map(Some);
                            }
                        }
                        // If there is matching then some group is matched.
                        unreachable!()
                    })
                }
                None => Err(Error::UnexepectedTokens(self.input.to_owned())),
            }
        }
    }

    /// Split the input into the vector of tokens.
    pub fn lex(mut self) -> Result<Vec<Token>> {
        let mut tokens = Vec::with_capacity(128);
        loop {
            let nxt = self.next()?;
            if let Some(nxt) = nxt {
                tokens.push(nxt);
            } else {
                break;
            }
        }
        Ok(tokens)
    }
}

#[cfg(test)]
mod tests {
    use {
        super::{Lexer, Result},
        crate::lexer::{ident, Keyword, Token::*},
    };

    #[test]
    fn test_lex() -> Result<()> {
        let input = "
void foo() {
    int a, b, c;
    a = 0;
    b=4; c=5;
    a=b+c;
}
";
        let tokens = Lexer::new(&input).lex()?;
        assert_eq!(
            tokens,
            vec![
                Keyword(Keyword::Void),
                ident("foo"),
                LPar,
                RPar,
                LCb,
                Keyword(Keyword::Int),
                ident("a"),
                Comma,
                ident("b"),
                Comma,
                ident("c"),
                Semicolon,
                ident("a"),
                Assign,
                Number(0),
                Semicolon,
                ident("b"),
                Assign,
                Number(4),
                Semicolon,
                ident("c"),
                Assign,
                Number(5),
                Semicolon,
                ident("a"),
                Assign,
                ident("b"),
                Plus,
                ident("c"),
                Semicolon,
                RCb,
            ]
        );

        Ok(())
    }
}

File: hs/app/Lexer.hs
module Lexer (Keyword(..), Token(..), nextToken, regex, lex0, Error(..), fromIdent) where

import Control.Applicative ((<|>))
import Data.List (findIndex)
import Data.Maybe (fromJust)
import Text.Read (readMaybe)
import Text.Regex.TDFA ((=~))

data Keyword
  = If
  | Else
  | Int
  | Unsigned
  | Void
  | Bool
  | Goto
  | Return
  deriving (Eq, Show)

-- Try to recognize a keyword.
makeKeyword :: String -> Maybe Keyword
makeKeyword a =
  case a of
    "if" -> Just If
    "else" -> Just Else
    "int" -> Just Int
    "unsigned" -> Just Unsigned
    "void" -> Just Void
    "bool" -> Just Bool
    "goto" -> Just Goto
    "return" -> Just Return
    _ -> Nothing

data Token
  = Eq
  | Gq
  | Lq
  | Shl
  | Shr
  | Tilda
  | LPar
  | RPar
  | LSq
  | RSq
  | LCb
  | RCb
  | Plus
  | Minus
  | Star
  | Slash
  | Percent
  | Assign
  | Gt
  | Lt
  | And
  | Xor
  | Or
  | Not
  | Comma
  | Semicolon
  | Colon
  | Keyword Keyword
  | Ident String
  | Number Int
  deriving (Eq, Show)

fromIdent :: Token -> String
fromIdent (Ident n) = n
fromIdent _ = error "expected an ident"

-- TODO: more informative.
data Error = Unexpected String | ParseError (Maybe [Token])
  deriving (Show)

{-
 - Match groups:
 - \* 0 - String without whitespaces
 - \* 1 - EOF
 - \* 2+ Tokens
-}
tokenRegexs :: [(String, String -> Maybe Token)]
tokenRegexs =
  [ ( "[0-9]+",
      fmap Number . readMaybe
    ),
    ( "[_a-zA-Z][_a-zA-Z0-9]*",
      -- Firstly try to extract keyword, then make an ident.
      \x -> fmap Keyword (makeKeyword x) <|> Just (Ident x)
    )
  ]
    ++ map
      (\(x, cons) -> (x, Just . const cons))
      primitiveTokens
  where
    primitiveTokens :: [(String, Token)]
    primitiveTokens =
      [ ("==", Eq),
        (">=", Gq),
        ("<=", Lq),
        ("<<", Shl),
        (">>", Shr),
        ("~", Tilda),
        ("\\(", LPar),
        ("\\)", RPar),
        ("\\[", LSq),
        ("\\]", RSq),
        ("\\{", LCb),
        ("\\}", RCb),
        ("\\+", Plus),
        ("\\-", Minus),
        ("\\*", Star),
        ("/", Slash),
        ("%", Percent),
        ("=", Assign),
        (">", Gt),
        ("<", Lt),
        ("&", And),
        ("\\^", Xor),
        ("\\|", Or),
        ("!", Not),
        (",", Comma),
        (";", Semicolon),
        (":", Colon)
      ]

-- Regexp that used to match tokens.
regex :: String
regex =
  "^[[:space:]]*(($)" ++ foldl f "" tokenRegexs ++ ")"
  where
    f s (re, _) = s ++ "|(" ++ re ++ ")"

-- Recognize a next token from the input.
nextToken :: String -> Either Error (String, Maybe Token)
nextToken s =
  if s == ""
    then return ("", Nothing)
    else
      let (_, m, suf, groups) = s =~ regex :: (String, String, String, [String])
       in if m == ""
            then
              -- No match at all.
              Left $ Unexpected s
            else
              let matched = head groups
                  -- As the string was matched => there is
                  -- some "token" group that was matched (or it is EOF).
                  --
                  -- Token groups started from 2, so take a suffix starting
                  -- from third element and try to find a match.
                  grp = (findIndex (/= "") $ (tail . tail) groups)
                  unwrapOrElse :: Maybe Token -> Either Error Token
                  unwrapOrElse mt = fromJust $ fmap return mt <|> Just (Left (Unexpected s))
               in case grp of
                    -- No token groups => EOF matched.
                    Nothing -> return ("", Nothing)
                    Just g -> do
                      -- Extract constructor for token by group.
                      let cons = fmap snd tokenRegexs !! g
                      -- Create a token from matched.
                      token <- unwrapOrElse $ cons matched
                      return (suf, Just token)

-- Lex input. In the case of error, returns a value that contains a suffix
-- where error token begins.
--
lex0 :: String -> Either Error [Token]
lex0 s = do
  (suf, token) <- nextToken s
  case token of
    Nothing -> return []
    Just tok -> do
      tokens <- lex0 suf
      return $ tok : tokens
